File: packages/takanawa-capacitor/ios/Sources/TakanawaCapacitorPlugin/TakanawaSnapshotMapper.swift
import Capacitor
import Foundation
import Takanawa

internal func snapshotPayload(_ snapshot: DownloadSnapshot, lastError: String? = nil) -> JSObject {
  var payload = JSObject()
  payload["phase"] = snapshot.phase.jsPhase
  payload["contentLen"] = String(snapshot.contentLen)
  payload["downloadedBytes"] = String(snapshot.downloadedBytes)
  payload["chunkSize"] = String(snapshot.chunkSize)
  payload["chunkCount"] = String(snapshot.chunkCount)
  payload["completedChunks"] = String(snapshot.completedChunks)
  payload["activeIo"] = snapshot.activeIo
  if let lastError, !lastError.isEmpty {
    payload["lastError"] = lastError
  }
  return payload
}

internal func speedPayload(_ snapshot: DownloadSpeedSnapshot) -> JSObject {
  var payload = JSObject()
  payload["phase"] = snapshot.phase.jsPhase
  payload["contentLen"] = String(snapshot.contentLen)
  payload["receivedBytes"] = String(snapshot.receivedBytes)
  payload["intervalBytes"] = String(snapshot.intervalBytes)
  payload["elapsedMillis"] = String(snapshot.elapsedMillis)
  payload["bytesPerSecond"] = snapshot.bytesPerSecond
  payload["activeIo"] = snapshot.activeIo
  return payload
}

internal extension DownloadPhase {
  var jsPhase: String {
    switch self {
    case .created:
      return "created"
    case .running:
      return "running"
    case .pausing:
      return "pausing"
    case .paused:
      return "paused"
    case .cancelling:
      return "cancelling"
    case .cancelled:
      return "cancelled"
    case .completed:
      return "completed"
    case .failed:
      return "failed"
    }
  }

  var isTerminal: Bool {
    self == .completed || self == .failed || self == .cancelled
  }
}

File: packages/takanawa-capacitor/android/src/main/kotlin/ai/yetanother/takanawa/capacitor/TakanawaSnapshotMapper.kt
package ai.yetanother.takanawa.capacitor

import ai.yetanother.takanawa.DownloadPhase
import ai.yetanother.takanawa.DownloadSnapshot
import com.getcapacitor.JSObject
import java.util.Locale

internal fun DownloadSnapshot.toJSObject(lastError: String? = null): JSObject {
    val payload = JSObject()
    payload.put("phase", phase.toJSPhase())
    payload.put("contentLen", contentLen.toString())
    payload.put("downloadedBytes", downloadedBytes.toString())
    payload.put("chunkSize", chunkSize.toString())
    payload.put("chunkCount", chunkCount.toString())
    payload.put("completedChunks", completedChunks.toString())
    payload.put("activeIo", activeIo)
    if (!lastError.isNullOrBlank()) {
        payload.put("lastError", lastError)
    }
    return payload
}

internal fun DownloadPhase.toJSPhase(): String = name.lowercase(Locale.US)

internal fun DownloadPhase.isTerminal(): Boolean =
    this == DownloadPhase.COMPLETED || this == DownloadPhase.FAILED || this == DownloadPhase.CANCELLED
